edx/edx-enterprise
# [ENT-125] Dynamically load course modes from enrollment API

**Description:**

Previously, the course mode box for enrolling users from the Enterprise app was limited to three "blessed" modes - audit, verified, and professional. However, it became clear that this was insufficient.

This PR modifies the view to dynamically load all available enrollment modes for a course directly from the Enrollment API, and presents them with the custom names given to them, rather than label them by internal value.

In addition, the other three valid modes (no-id-professional, credit, and honor) have been added to the allowable options, meaning that in cases where Javascript is disabled, all valid modes can be chosen correctly.

**Screenshots:**

No course ID entered:

![screen shot 2017-01-04 at 9 22 26 pm](https://cloud.githubusercontent.com/assets/7773758/21667301/b32f9480-d2c5-11e6-8ff4-157242817f34.png)

Valid course ID entered:

![screen shot 2017-01-04 at 9 23 42 pm](https://cloud.githubusercontent.com/assets/7773758/21667300/b32f3ee0-d2c5-11e6-9789-044353e83e2a.png)

Invalid course ID entered:

![screen shot 2017-01-04 at 9 24 05 pm](https://cloud.githubusercontent.com/assets/7773758/21667302/b330cd5a-d2c5-11e6-92e8-22a093ecdeea.png)

Javascript disabled:

![screen shot 2017-01-04 at 9 25 52 pm](https://cloud.githubusercontent.com/assets/7773758/21667299/b32b7ecc-d2c5-11e6-8bd6-793beba5f882.png)


**JIRA:** https://openedx.atlassian.net/browse/ENT-125

**Dependencies:** None

**Merge deadline:** 2017-01-17

**Sandbox**: https://ent91.opencraft.hosting (built-in `staff` user)

**Testing instructions:**

1. Create a couple custom course modes for a given course; use distinctive names.
2. Navigate to the Manage Learners view of the Enterprise Admin.
    * Observe that the course mode dropdown is disabled, and is filled with a prompt to enter a valid course ID.
2. Enter the course ID of the course you've chosen and tab away from the course box.
    * Observe that the course mode dropdown is populated with the course modes you created.
3. Select one of the modes and submit the form.
    * Observe that although the form has errors, the course mode dropdown is returned to the same state and position it was in prior to submitting the form.
4. Fill in the email address of an existing user not yet linked with this EnterpriseCustomer, and submit the form again.
    * Observe that the user is now enrolled in the given course, with the given course mode you chose from the dropdown.
3. Enter an invalid course ID in the course ID box.
    * Observe that the course mode box returns to its original, disabled, state.
4. Disable Javascript and refresh the page.
    * Observe that although nothing is dynamically loaded, all six valid course mode options are presented for you to select from.

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.

**Author concerns:**

1. I have some concerns about a11y with this PR. Multiple browsers (I tested Safari and Firefox) appear to choose tab order prior to calling the `.blur()` event on the control that's being departed. As a result, if the mode drop-down is disabled prior to tabbing away from the course ID box, then the focus will go to the submit button, rather than the dropdown, even if the AJAX call will enable the dropdown. There are a couple ways around this; none pretty:
    * Whenever course mode items are successfully loaded, set the focus to the dropdown. This is ugly because it prevents the user from choosing whatever focus item they like.
    * Don't actually disable the dropdown when there aren't valid options. This isn't ideal, but doesn't cause any harm, because when it's in such a state, there's only one option to choose from anyway.
